cub/cub/device/dispatch/kernels/kernel_scan_warpspeed.cuh

```diff
@@ -0,0 +1,801 @@
+      // Bitmask with 1 bits indicating which lane has an XX state.
+      int warpIsEmpty   = __reduce_or_sync(~0, laneIsEmpty);
+      int warpIsCumSum  = __reduce_or_sync(~0, laneIsCumSum);
+      int warpIsPrivSum = __reduce_or_sync(~0, laneIsPrivSum);
+
+      if (warpIsEmpty != 0)
+      {
+        break;
+      }
+      // Now we have either all private sums, or a mix of private sums and
+      // cumulative sums.
+
+      // Bitmask with a 1 bit indicating the position of the right-most
+      // CUM_SUM state. If no CUM_SUM state present, value is zero.
+      int warpRightMostCumSum = warpIsCumSum & -warpIsCumSum;
+      // Bitmask with 1 bits to the right of the right-most CUM_SUM state.
+      // If no CUM_SUM state present, value is all ones.
+      int maskRightOfCumSum = warpRightMostCumSum - 1;
+
+      // Sum all values of lanes containing either
+      // (a) the right-most CUM_SUM, or
+      // (b) subsequent PRIV_SUMs.
+      AccumT localSum{};
+      int maskSumParticipants = warpRightMostCumSum | maskRightOfCumSum;
+
+      if ((maskSumParticipants & lanemaskEq) != 0)
+      {
+        localSum = regTmpStates[idx].value;
+      }
+      localSum = warpReduce(localSum, scan_op);
+
+      if (warpIsCumSum == 0)
+      {
+        sumExclusiveCtaCur = scan_op(sumExclusiveCtaCur, localSum);
+      }
+      else
+      {
+        sumExclusiveCtaCur = localSum;
+      }
+      // idxTileCur can go beyond idxTileNext.
+      idxTileCur += __popc(maskSumParticipants);
+    }
+  }
+
+  // We are not storing CUM_SUM states, because it makes updating idxTileCur
+  // difficult. Storing CUM_SUM states is the more robust approach though: if
+  // for whatever reason GETNEXTWORKID fails at some point, any freshly launched
+  // block must look back many iterations to figure out the current
+  // sumExclusiveCta.
+  //
+  // Just for future reference, the following commented code is kept around.
+
+  // if (idxTileNext > 0) {
+  //   if (specialRegisters.laneIdx == 0) {
+  //     storeLookback(ptrTmpBuffer, idxTileNext - 1, CUM_SUM, sumExclusiveCtaCur);
+  //   }
+  // }
+
+  return sumExclusiveCtaCur;
+}
+
+namespace ptx = cuda::ptx;
+
+template <typename InputT, typename OutputT, typename AccumT>
+struct scanKernelParams
+{
+  InputT* ptrIn;
+  OutputT* ptrOut;
+  tmp_state_t<AccumT>* ptrTmp;
+  size_t numElem;
+  int numStages;
+};
+// Definition of squads
+
+_CCCL_GLOBAL_CONSTANT SquadDesc squadReduce{/*squadIdx=*/0, /*numWarps=*/4};
+_CCCL_GLOBAL_CONSTANT SquadDesc squadScanStore{/*squadIdx=*/1, /*numWarps=*/4};
+_CCCL_GLOBAL_CONSTANT SquadDesc squadLoad{/*squadIdx=*/2, /*numWarps=*/1};
+_CCCL_GLOBAL_CONSTANT SquadDesc squadSched{/*squadIdx=*/3, /*numWarps=*/1};
+_CCCL_GLOBAL_CONSTANT SquadDesc squadLookback{/*squadIdx=*/4, /*numWarps=*/1};
+
+_CCCL_GLOBAL_CONSTANT SquadDesc scanSquads[] = {
+  squadReduce,
+  squadScanStore,
+  squadLoad,
+  squadSched,
+  squadLookback,
+};
```

Comment:
Nitpick: I believe this is a tad error prone, I would prefer if we would have a utility function that creates a set of `SquadDesc` with the given number of warps and then have named references to the elements of the array

Reply:
Interesting. How would that look like?